### PR TITLE
Fixes tests that was broken with enhancements made to the `lexer`

### PR DESCRIPTION
**Pull request by sogko**
Hi @Matthias247 

_The changes here includes changes in PR #1 (which updatesmatthias247/graphql:lexperf with latest changes from graphql-go/graphql:master). Merge that first before reviewing the changes here. 👍_

---

Fixes tests that was broken with enhancements made to the `lexer` with commit https://github.com/Matthias247/graphql/commit/be39d41a18a9934094987a110403ecdab4a9de86
- Running `go test` would have passed previously, but running all tests with `go test ./...` would fail.
- Increased coverage for `lexer` package to 100%
- Added more tests to cover parsing unicode strings as well. (see graphql-go/graphql#135)

This improvements made to the `lexer`  in this branch do improve the perf by quite a bit (See separate discussion at https://github.com/sogko/graphql/pull/22#issuecomment-225395914). Would be glad to add it to v0.5.0 release.

Cheers!

### CODE REVIEW

**Comment by Matthias247** [comment]
Hi,
sorry, I don't understand the intent of these pull requests? Can't you directly merge my PR into the main go-graphql repo? Or merge my commit into one of your PRs? I only forked this repo temporarily in order to provide the pull request. And unfortunatly I'm currently not able to review all other changes.

**Comment by sogko** [comment]
Hi @Matthias247 

Oops, let me explain :)

Your `lexperf` branch got outdated (graphql-go/graphql#137) after the main repo merged huge changes from graphql-go/graphql#123.

The only relevant commit in this PR is https://github.com/Matthias247/graphql/pull/2/commits/2762599d53689e8316b5693c27e0db855ad1ce8f  (I've fixed the `lexer` tests after your improvements to the `lexer` and fixed some issues with your original implementation).

The other commits would be gone after you merge in PR #1 (which will bring your `lexperf` branch up-to-date, you don't have to review those in #1 if you don't want to :) )

After both PR #1 and #2 has been merged, I can close your original PR graphql-graphql-go#137

Hope that clears it up a little bit?

Cheers!

**Comment by Matthias247** [comment]
Ah, I see it now. But wouldn't it make much more sense to rebase the PR on current graphql-go/master and merge it then?  I can do that. Merging the things here and merging the merged product back in the original repo sounds like duplicated merge of your commits.

**Comment by Matthias247** [comment]
As an alternative you could also cherry pick my single lexer commit.

**Comment by sogko** [comment]
Yeah thats the ideal way, please do that :) 
I was just got too excited with bringing in your changes in and getting the tests fixed.

**Comment by Matthias247** [comment]
@sogko Rebased my PR on top of the current master branch. However that no misses the latest changes to the lexer that you made in the last commit here (e.g. fixing tests).

File: language/lexer/lexer.go [hunk 1]
```diff
@@ -106,24 +102,29 @@ func Lex(s *source.Source) Lexer {
...
 	bodyLength := len(body)
-	end := position + 1
+	endByte := position + 1
+	endRune := runePosition + 1
 	for {
-		code, n := runeAt(body, end)
-		if (end != bodyLength) &&
+		code, n := runeAt(body, endByte)
+		if (endByte != bodyLength) &&
+			n == 1 && // only ASCII chars are allowed to be names
```
Comment (Matthias247) [inline language/lexer/lexer.go]: The n == 1 check is superfluos here, since the next and condition only checks for char codes which are all single-byte characters.

File: language/lexer/lexer.go [hunk 2]
```diff
@@ -237,6 +240,7 @@ func readString(s *source.Source, start int) (Token, error) {
 					valueBuffer.WriteRune('"')
 					break
 				case '/':
+					valueBuffer.WriteRune('\\')
```
Comment (Matthias247) [inline language/lexer/lexer.go]: What's the reason for adding this?

File: language/lexer/lexer.go (hunk 1 above)
Comment (sogko) [inline language/lexer/lexer.go]: Yeah, I agree, I'll remove that 👍

**Comment by Matthias247** [comment]
I have seen that you added functionality to also keep the current runeIndex in memory. I think that's ok. But I don't know if passing that to `NewSyntaxError` is the correct thing. If I understand the method and `GetLocation` correctly they are operating on bytes (the lineRegexp does). If showing the correct column (as a rune and not byte index) inside the line is imporant then I think the amount of runes up to `position` can be counted for the line inside the GetLocation function.

File: language/lexer/lexer.go (hunk 2 above)
Comment (sogko) [inline language/lexer/lexer.go]: Hmm, this was earlier added to satisfy the test for `TestLexer_LexesStrings`

```
        {
            Body: "\"slashes \\\\ \\/\"",
            Expected: Token{
                Kind:  TokenKind[STRING],
                Start: 0,
                End:   15,
                Value: "slashes \\ \\/",
            },
        },


```

(previously `valueBuffer ...` was `value += "\\/"`)

Turns out the test was wrong, it should be:

```
        {
            Body: "\"slashes \\\\ \\/\"",
            Expected: Token{
                Kind:  TokenKind[STRING],
                Start: 0,
                End:   15,
                Value: "slashes \\ /",
            },
        },
```

(Reference: https://github.com/graphql/graphql-js/blob/master/src/language/__tests__/lexer-test.js#L142)

I'll make the changes and push it back to your branch, nice catch! 👍

Comment (Matthias247) [inline language/lexer/lexer.go]: Hmm, that double escaping is tricky to see and get right.
I think you might use raw string literals in backticks to make it a little easier to understand :)
Like

```
Body: `"slashes \\ \/"`,
```

However I just ask myself why and if `/` is escaped at all.

**Comment by sogko** [comment]
Yeah, I'm using the `runePosition` for `SyntaxError`, which gives out a nicely formatted error and it highlights the position of rune at which the syntax error occurs. (Since `readString()` can contain unicode text)

It certainly has limitations at the moment (for e.g. runes that are 3- and 4-bytes wide may not sit in one column, resulting the `^` "pointer" being off.)

I didn't spend much time on that, it certainly would be nice to see this getting tackled. Do you want to address that in a separate PR? It would be related to ensuring a nice coverage over unicode support. (Current test suite are lacking that at the moment)

**Comment by Matthias247** [comment]
I think the most reasonable way to do this is to track the current line as well as in the Lexer to avoid the GetLocation call alltogether. The current implementation of GetLocation is quite costly, as it allocates a new string for each line. It could be replaced by a linear scan through the body to scan for line breaks and count columns. But as this is basically already done during lexing it would be redundant, especially if there are multiple errors.

So the lexer would need to track multiple variables:
- Current byte offset
- Current line nr
- Current column nr

Sorry - don't have too much time to look into this now. For me the most imporant thing was to optimize the "happy path".

**Comment by Matthias247** [comment]
So how should be proceed with this? Was your plan that I merge this and you merge the result back into graphql repo? If yes then this is PR refers the wrong branch, because my rebased changes are in the lexperf branch and not on the master branch. Otherwise you can simply merge my PR to the main graphql repo. And then merge your changes on top of it.

**Comment by sogko** [comment]
Haha sure no problem, I'll put the improvement for `GetLocation()` in the backlog. Right now we are focused on getting it right and relying on the test suites to ensure that we are on the right track.

I'll merge your PR to the main repo and push the updates on top of it 👌

Thanks for contributing your time on this, appreciate it very much!

Cheers!
